pkg/common/queue/chunkqueue.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+)
+
+const (
+	defaultSizePerChunk = 1024
+)
+
+// ChunkQueue is a generic, efficient and GC-friendly queue
+type ChunkQueue[T any] struct {
+	// [head, tail] is the section of chunks in use
+	head int
+	tail int
+	size int
+
+	chunks       []*chunk[T]
+	chunkSize    int
+	chunkPool    sync.Pool
+	defaultValue T
+}
+
+func (q *ChunkQueue[T]) headChunk() *chunk[T] {
+	return q.chunks[q.head]
+}
+
+func (q *ChunkQueue[T]) tailChunk() *chunk[T] {
+	return q.chunks[q.tail]
+}
+
+// NewChunkQueue creates a new ChunkQueue
+func NewChunkQueue[T any]() *ChunkQueue[T] {
+	return NewChunkQueueLeastCapacity[T](1)
+}
+
+// NewChunkQueueLeastCapacity creates a ChunkQueue with an argument minCapacity.
+// It requests that the queue capacity be at least minCapacity. And it's similar
+// to the cap argument when making a slice using make([]T, len, cap)
+func NewChunkQueueLeastCapacity[T any](minCapacity int) *ChunkQueue[T] {
+	if unsafe.Sizeof(*new(T)) == 0 {
+		log.Panic("Cannot create queue of type that size == 0")
+	}
+
+	chunkSize := defaultSizePerChunk / int(unsafe.Sizeof(*new(T)))
+	if chunkSize < 16 {
+		chunkSize = 16
+	}
+
+	q := &ChunkQueue[T]{
+		head:      0,
+		tail:      -1, // temporary be negative
+		size:      0,
+		chunkSize: chunkSize,
+	}
+	q.chunkPool = sync.Pool{
+		New: func() any {
+			return newChunk[T](q.chunkSize, q)
+		},
+	}
+
+	q.chunks = make([]*chunk[T], 32, 32)
+	q.expandSpace(minCapacity)
+	return q
+}
+
+// Size() returns the number of elements in queue
+func (q *ChunkQueue[T]) Size() int {
+	return q.size
+}
+
+// Cap() returns the capacity of the queue. The queue can hold more elements
+// than that number by automatic expansion
+func (q *ChunkQueue[T]) Cap() int {
+	return q.chunkSize*(q.tail-q.head) - q.chunks[q.head].l - 1
```

Comment:
What if a `q.tail == -1` and `q.head == 0 `?

Reply:
It never happens. It's only the case that `q.tail` would temporarily be negative when the queue has not been initiated and returned, since `q.tail` will be corrected in `(q *ChunkQueue[T]) expandSpace()`. However, the logic here is too tricky and I'll rewrite this part.